### PR TITLE
build: pre-commit infrastructure and update travis

### PR DESCRIPTION
**Pull request by yangchoo**
Add pre-commit infrastructure to help with basic lint tasks. This is in
preparation of adding an autoformatter for the repo.

As part of this, the travis file is reorganized to help with clearer
delineation of jobs.

Documention is also updated, with the Contributing section updated with
more explicit testing steps and the new environment and lint setup.

This includes all the fixes from the pre-commit hooks added.

### CODE REVIEW

**Comment by yangchoo** [comment]
Please do check out the Travis build for the new test matrix: https://travis-ci.org/github/autoprotocol/autoprotocol-python/builds/689570208

File: .travis.yml [hunk 1]
```diff
@@ -5,12 +5,25 @@ env:
...
-install: pip install tox-travis
-script: tox
+jobs:
+  include:
+  - python: '3.6'
+    env: TOXENV=docs
+  - python: '3.6'
+    env: TOXENV=lint
+  - python: '3.6'
+    env: TOXENV=clean,py36,stats
```
Comment (rhysormond) [inline .travis.yml:15]: is the clean step here necessary still?

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -11,6 +11,58 @@ Features and Bugs
...
+Python Virtual Environment (Optional):
+    Use of virtual environment to isolate development environment is
+    highly recommended. There are several tools available such as
+    conda_ and pyenv_.
+
+Dependencies:
+    We recommend first activating your desired virtualenv, then
+    installing the dependencies using the snippet below.
+
+.. code-block:: bash
```
Comment (rhysormond) [inline CONTRIBUTING.rst:29]: extremely minor nit: these aren't really using anything bash-specific and could just be `sh` which *appears* to be a supported tag

**Comment by rhysormond** [comment]
somewhat related, can we configure travis so that it's using something newer than xenial?

File: .travis.yml (hunk 1 above)
Comment (yangchoo) [reply .travis.yml:15]: ah, yea, its to blow away coverage artifacts if present. It seems common in other travis integration libraries, so I assume its doing something but haven't validated

File: CONTRIBUTING.rst (hunk 2 above)
Comment (yangchoo) [reply CONTRIBUTING.rst:29]: cool! will do

Comment (yangchoo) [reply CONTRIBUTING.rst:29]: yup sh works

**Comment by yangchoo** [comment]
yea, could use Bionic (https://docs.travis-ci.com/user/reference/linux/).

It shouldn't really matter since we're using mostly pip/tox for the running, so it should be fine to upgrade. I'll try it

**Pull request merged**